### ISSUE TITLE
Estimating Gradients for Stochastic Neurons

### ISSUE THREAD

**Issue by crosleythomas** [fchollet/keras#4903]
Are there plans/interest in incorporating training ability for the use of stochastic units in hidden layers? 

For example, some of the methods described in https://arxiv.org/pdf/1308.3432.pdf or the original REINFORCE paper.

I think this would be a very useful addition for the rising popularity of variational methods.  As far as I can tell there is currently no support for training a custom layer that has stochastic neuron, for example binomial neurons that are drawn from a previous layer with sigmoid outputs.

**Comment by bstriner** [comment]
You can make custom stochastic layers in Keras but I don't know if anyone has implemented that particular model yet. I made a layer for stochastic weight sampling and variational regularization, which might give you some ideas of how to implement it. 

https://github.com/bstriner/bayesian_dense

You could also look through the code for the probabilistic relu in advanced activations.

Bengio is on the paper you posted, so I don't think anyone will argue with trying to add it to main line keras. Anyone else up for testing this out? Doesn't seem like the hardest implementation and should be really useful.

Cheers,
Ben

**Comment by crosleythomas** [comment]
Thanks for the link.

If I understand correctly it looks like for the bayesian dense approach the stochasticity is not parameterized by weights in the network.  So, like variational auto-encoders, which use noise as an auxiliary input, the whole network is still differentiable.  

Gradient estimation is needed when the output of neurons in hidden layers are stochastic because the random sampling operation is generally not differentiable.  For example, if you have a sigmoid activation function in a layer that feeds into a layer that does binomial sampling, the binomial sampling layer is not differentiable and the weights leading into the sigmoid will not be trained.   Tang and Salakhutdinov's paper on Stochastic Feed-Forward Neural Networks is one example of a network that uses stochastic activations for layer outputs. https://papers.nips.cc/paper/5026-learning-stochastic-feedforward-neural-networks.pdf 

To do this I think we would have to override the gradient computation in the backprop steps.  The best way to do this may be having a set of stochastic layers that each take an estimator as an argument.

Also, I'm not seeing a probabilistic relu in the advanced activations (I see LeakyReLU, PReLU, ELU, ThresholdedReLU, and SReLU - none of which are probabilistic).

**Comment by bstriner** [comment]
There are several approaches in the Bengio paper that didn't seem to need overriding the gradient computation. The noisy rectifier and STS unit both seem pretty straightforward. Sections 3 and 4 might require custom operations to get the backprop working correctly.
